kani-driver/src/args/playback_args.rs

```diff
@@ -0,0 +1,154 @@
+use clap::{Error, Parser, ValueEnum};
+use std::path::PathBuf;
+
+/// Execute concrete playback testcases of a local package.
+#[derive(Debug, Parser)]
+pub struct CargoPlaybackArgs {
+    #[command(flatten)]
+    pub playback: PlaybackArgs,
+
+    /// Arguments to pass down to Cargo
+    #[command(flatten)]
+    pub cargo: CargoArgs,
+}
+
+/// Execute concrete playback testcases of a local crate.
+#[derive(Debug, Parser)]
+pub struct KaniPlaybackArgs {
+    /// Rust crate's top file location.
+    pub input: PathBuf,
+
+    #[command(flatten)]
+    pub playback: PlaybackArgs,
+}
+
+/// Playback subcommand arguments.
+#[derive(Debug, clap::Args)]
+pub struct PlaybackArgs {
+    /// Common args always available to Kani subcommands.
+    #[command(flatten)]
+    pub common_opts: CommonArgs,
+
+    /// Specify which test will be executed by the playback args.
+    #[arg(long)]
+    pub test: String,
+
+    /// Compile but don't actually run the tests.
+    #[arg(long)]
+    pub only_codegen: bool,
+
+    // TODO: We should make this a common option to all subcommands.
+    /// Control the subcommand output.
+    #[arg(long, default_value = "human")]
+    pub message_format: MessageFormat,
+}
+
+/// Message formats available for the subcommand.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, ValueEnum, strum_macros::Display)]
+#[strum(serialize_all = "kebab-case")]
+pub enum MessageFormat {
+    /// Print diagnostic messages in a user friendly format.
+    Human,
+    /// Print diagnostic messages in JSON format.
+    Json,
+}
+
+impl ValidateArgs for CargoPlaybackArgs {
+    fn validate(&self) -> Result<(), Error> {
+        self.playback.validate()?;
+        self.cargo.validate()
+    }
+}
+
+impl ValidateArgs for KaniPlaybackArgs {
+    fn validate(&self) -> Result<(), Error> {
+        self.playback.validate()?;
+        if !self.input.is_file() {
+            return Err(Error::raw(
+                ErrorKind::InvalidValue,
+                &format!(
+                    "Invalid argument: Input invalid. `{}` is not a regular file.",
+                    self.input.display()
+                ),
+            ));
+        }
+        Ok(())
+    }
+}
+
+impl ValidateArgs for PlaybackArgs {
+    fn validate(&self) -> Result<(), Error> {
+        self.common_opts.validate()?;
+        if !self.common_opts.unstable_features.contains(&UnstableFeatures::ConcretePlayback) {
+            return Err(Error::raw(
+                ErrorKind::MissingRequiredArgument,
+                "The `playback` subcommand is unstable and requires `-Z concrete-playback` \
+                to be used.",
+            ));
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use clap::Parser;
+
+    use super::*;
+
+    #[test]
+    fn check_cargo_parse_test_works() {
+        let input = "playback -Z concrete-playback --test TEST_NAME".split_whitespace();
+        let args = CargoPlaybackArgs::try_parse_from(input.clone()).unwrap();
```

Comment:
Is the clone needed?

Reply:
Probably not. I initially had one test for both